### DISCUSSION TITLE
I need help with GGML with GPU Accelleration. Nothing I do resolves issue!

### DISCUSSION THREAD

**Discussion by viperwasp** [oobabooga/text-generation-webui#2635] [Q&A] [answered]
Trying to figure out what dependencies I need? Please if possible list the known dependancies and version your using if you can confirm a working version. 

I have tried Visual Studios 2022 Community and then switched to 2019 Builds Version as someone recommend. 

I downloaded Microsoft Windows SDK 10.0.20348.1

I had Nvidia Cuda Toolkit 12.1 but I downgraded to 11.7 as I got an error in the Ooba Windows updater when building
the Cuda wheel I think. 11.7 fixed that issue. 

I updated Oobabooga. I updated Nvidia Drivers, I have a RTX 4080, Windows 10 Home. 
I tried using the GPU Acceleration setup guide
https://github.com/oobabooga/text-generation-webui/blob/main/docs/llama.cpp-models.md

I downloaded Ooba from this link
https://github.com/oobabooga/text-generation-webui/releases/download/installers/oobabooga_windows.zip
located https://github.com/oobabooga/text-generation-webui just encase I'm using the wrong fork or oneclickinstaller. 

I followed these steps. I even tried all three backends. I also tried following the steps here
https://github.com/oobabooga/text-generation-webui/discussions/2488

Quote
mykeehu
Jun 2, 2023
Author
Ok, now I was able to install GPU mode with these steps:

    I uninstalled all Visual Studio and installed [VS 2019 BuildTools](https://my.visualstudio.com/Downloads?q=visual%20studio%202019&wt.mc_id=o~msft~vscom~older-downloads)
    add to PATH this line: C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64 
    I had to install the latest [Windows 10 SDK](https://developer.microsoft.com/en-us/windows/downloads/sdk-archive/) separately (not enough VS version!), including the ones called Desktop C++
    I had to copy four files from this folder: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.7\extras\visual_studio_integration\MSBuildExtensions to this folder: C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\MSBuild\Microsoft\VC\v160\BuildCustomizations
    then I ran the four commands above.
    I'm going to try it :)
    

At no point have I been able to get GGML to load into video memory. And the BLAS = 0 has never changed to a 1. 
I am almost completely out of ideas. I don't notice any strange errors etc. As for the "Ooba" settings I have tried a lot of settings. And saving/reloading the model. Restarting PC etc. In the link I provided above that has screenshots of what settings to choose in ooba like N GPU slider etc. I tried all of that. I even copied the four files over from the Nvidia Cuda folder to the 2019 Visual Studios folder like how mykeehu suggested. 
    
GGML models work fine on CPU/RAM only but no GPU Acceleration? Thanks. I have almost given up.

**Comment by DavidInRacine** [comment]
I am not an expert either, but I am pretty sure you need the GPTQ files instead of GGML -- which is the CPU only version. I don't have NVIDIA hardware so I must use GGML.

Good luck, and let us know how it goes.

**Reply by viperwasp** [reply]
No GGML is correct. That much I am sure GPTQ is for the groupsize and Wbit ones. GGML are the ones that go like 4_0 4_1, 5_0, 5_1 they run on llama.cpp and is it technically supported on Windows with GPU acceleration.

**Comment by ossirytk** [comment]
GGML is correct. You load a part of the model to cpu ram and part is handled with gpu vram. Ram is cheap, so it's easy to have 64gb ram and load 30B models. But cpu model loading and inference are slow, so you want to push as much work on the gpu as possible. I have 64gb of ram and can easily load large models, but the load time for the model to be loaded into memory is really long. Inference is also really slow. I only have 8gb vram trash gpu, but the performance is doubled for 13B model when gpu is enabled. I get like 28ms/token.

**Reply by viperwasp** [reply]
Yes so far I run 13B completely on GPU which is of course extremely fast. I also run GGML 30B and GGML 65B completely on RAM. I also have 64GB. 30B is fast enough but nothing amazing. 65B is slow but if I'm doing other things it's fine in a limited capacity but I would love to offload some to my GPU to increase the speed.

**Comment by jparmstr** [comment]
Maybe try using Llama.cpp directly. This might narrow down whether it's a WebUI bug or something else. Llama.cpp now provides precompiled binaries, so it's easy to test. [Latest release, CUDA 11.7](https://github.com/ggerganov/llama.cpp/releases/download/master-e4caa8d/llama-master-e4caa8d-bin-win-cublas-cu11.7.1-x64.zip).

Here is a test command you could try:
`.\main.exe --model llama-7b.ggmlv3.q4_0.bin -p "Instruction: Write a short story about Llamas. Response: " -n 200 -ngl 32`

Provide the full path to your model if it isn't in the same folder as Llama.cpp's main.exe. Here is the beginning of the output from successful GPU offloading:

```
main: build = 664 (e4caa8d)
main: seed  = 1686600076
ggml_init_cublas: found 1 CUDA devices:
  Device 0: NVIDIA GeForce RTX 3060 Ti
llama.cpp: loading model from ...\oobabooga_windows\text-generation-webui\models\llama-7b.ggmlv3.q4_0.bin
llama_model_load_internal: format     = ggjt v3 (latest)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 512
llama_model_load_internal: n_embd     = 4096
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 32
llama_model_load_internal: n_layer    = 32
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: n_ff       = 11008
llama_model_load_internal: n_parts    = 1
llama_model_load_internal: model size = 7B
llama_model_load_internal: ggml ctx size =    0.07 MB
llama_model_load_internal: using CUDA for GPU acceleration
llama_model_load_internal: mem required  = 1932.71 MB (+ 1026.00 MB per state)
llama_model_load_internal: allocating batch_size x 1 MB = 512 MB VRAM for the scratch buffer
llama_model_load_internal: offloading 32 layers to GPU
llama_model_load_internal: total VRAM used: 3987 MB
...................................................................................................
llama_init_from_file: kv self size  =  256.00 MB

system_info: n_threads = 3 / 6 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | VSX = 0 |
sampling: repeat_last_n = 64, repeat_penalty = 1.100000, presence_penalty = 0.000000, frequency_penalty = 0.000000, top_k = 40, tfs_z = 1.000000, top_p = 0.950000, typical_p = 1.000000, temp = 0.800000, mirostat = 0, mirostat_lr = 0.100000, mirostat_ent = 5.000000
generate: n_ctx = 512, n_batch = 512, n_predict = 300, n_keep = 0
```

If this fails, hopefully it will give you a more informative error message. If it works, you know that the problem is with Text Generation WebUI, not your CUDA setup.

**Reply by viperwasp** [reply]
Good idea I will try this later. Thanks.

**Reply by ossirytk** [reply]
I agree with this idea. If you have an issue with the vscode or cuda side, then the cmake for building llama.cpp will fail and you will get some error messages that might be helpful. If the build passes and you can run everything in terminal, then the issue is somewhere in python pip side.

**Reply by viperwasp** [reply]
**I did not follow the exact steps suggested. Instead I went and downloaded llama.cpp with cublas-cu12.1** 

I ran the following .bat file

title llama.cpp
:start
main -i --interactive-first -r "### Human:" --temp 0 -c 2048 -n -1 --ignore-eos --repeat_penalty 1.2 --instruct -n 200 -ngl 38 -m Wizard-Vicuna-30B-Uncensored.ggmlv3.q4_0.bin -t 8
pause
goto start

Here are the results

C:\AI\llama-master-e4caa8d-bin-win-cublas-cu12.1.0-x64>title llama.cpp

C\AI\llama-master-e4caa8d-bin-win-cublas-cu12.1.0-x64>main -i --interactive-first -r "### Human:" --temp 0 -c 2048 -n -1 --ignore-eos --repeat_penalty 1.2 --instruct -n 200 -ngl 38 -m Wizard-Vicuna-30B-Uncensored.ggmlv3.q4_0.bin -t 8
main: build = 664 (e4caa8d)
main: seed  = 1686640223
ggml_init_cublas: found 1 CUDA devices:
  Device 0: NVIDIA GeForce RTX 4080
llama.cpp: loading model from Wizard-Vicuna-30B-Uncensored.ggmlv3.q4_0.bin
llama_model_load_internal: format     = ggjt v3 (latest)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 2048
llama_model_load_internal: n_embd     = 6656
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 52
llama_model_load_internal: n_layer    = 60
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: n_ff       = 17920
llama_model_load_internal: n_parts    = 1
llama_model_load_internal: model size = 30B
llama_model_load_internal: ggml ctx size =    0.13 MB
llama_model_load_internal: using CUDA for GPU acceleration
llama_model_load_internal: mem required  = 8848.13 MB (+ 3124.00 MB per state)
llama_model_load_internal: allocating batch_size x 1 MB = 512 MB VRAM for the scratch buffer
llama_model_load_internal: offloading 38 layers to GPU
llama_model_load_internal: total VRAM used: 11421 MB
....................................................................................................
llama_init_from_file: kv self size  = 3120.00 MB

system_info: n_threads = 8 / 20 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | VSX = 0 |
main: interactive mode on.
Reverse prompt: '### Human:'
Reverse prompt: '### Instruction:

'
sampling: repeat_last_n = 64, repeat_penalty = 1.200000, presence_penalty = 0.000000, frequency_penalty = 0.000000, top_k = 40, tfs_z = 1.000000, top_p = 0.950000, typical_p = 1.000000, temp = 0.000000, mirostat = 0, mirostat_lr = 0.100000, mirostat_ent = 5.000000
generate: n_ctx = 2048, n_batch = 512, n_predict = 200, n_keep = 2


== Running in interactive mode. ==
 - Press Ctrl+C to interject at any time.
 - Press Return to return control to LLaMa.
 - To return control without starting a new line, end your input with '/'.
 - If you want to submit another line, end your input with '\'.
 
 
 **And yes it's able to write stories and it's very fast.** 
 
 1. Does this prove I setup everything fine with Visual Studios and Nvidia Cuda? I did 2022 community with latest 12.1 for cuda. 
 2. Any ideas why it won't work in "Ooba?" 
 
 I tried CMD and Powershell. 
 Ran exactly 
 
 pip uninstall -y llama-cpp-python
set CMAKE_ARGS=-DLLAMA_CUBLAS=on
set FORCE_CMAKE=1
pip install llama-cpp-python --no-cache-dir

It seemed to do stuff. I even tried it a few times. 
Now my exact setup order was 1. Visual Studios, 2. Nvidia toolkit, and then 3. Ooba. I did the windows setup with ooba via one click install. Copied the 30B model into the models folder. Make sure ooba was working. And then I ran those commands after closing ooba down. 

**Here is what I get in Ooba** 


bin C:\AI\oobabooga_windows2\installer_files\env\lib\site-packages\bitsandbytes\libbitsandbytes_cuda117.dll
2023-06-13 03:21:31 INFO:Loading Wizard-Vicuna-30B-Uncensored.ggmlv3.q4_0.bin...
2023-06-13 03:21:31 INFO:llama.cpp weights detected: models\Wizard-Vicuna-30B-Uncensored.ggmlv3.q4_0.bin

2023-06-13 03:21:31 INFO:Cache capacity is 0 bytes
llama.cpp: loading model from models\Wizard-Vicuna-30B-Uncensored.ggmlv3.q4_0.bin
llama_model_load_internal: format     = ggjt v3 (latest)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 2048
llama_model_load_internal: n_embd     = 6656
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 52
llama_model_load_internal: n_layer    = 60
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: n_ff       = 17920
llama_model_load_internal: n_parts    = 1
llama_model_load_internal: model size = 30B
llama_model_load_internal: ggml ctx size =    0.13 MB
llama_model_load_internal: mem required  = 19756.66 MB (+ 3124.00 MB per state)
.
llama_init_from_file: kv self size  = 3120.00 MB
AVX = 1 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 0 | SSE3 = 1 | VSX = 0 |
2023-06-13 03:21:31 INFO:Loaded the model in 0.05 seconds.

2023-06-13 03:21:31 INFO:Loading the extension "gallery"...
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.

**Possible ideas or solutions please let me know?** 
Thanks

**Reply by viperwasp** [reply]
I don't know if this means anything or not? But I just did pip uninstall -y llama-cpp-python
and it uninstalled. 

I did not reinstalled it. "Ooba" still works for my GGML.bin files? Do I have some other instance of of llama installed that it's drawing on? OR what is going on?

**Reply by viperwasp** [reply]
What about Pytorch do I need a specific version that supports Cuda or something like that? I tried a lot of different versions run into errors like 

RROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
xformers 0.0.20 requires torch==2.0.1, but you have torch 2.1.0.dev20230612+cu121 which is incompatible.

Regardless nothing has worked. lol

**Reply by viperwasp** [reply]
I also noticed that...
Cuda toolkit in in my System Variables 3 times? 

Once as Variable name Cuda_Path and then also as Cuda_path_v12_1 
Finally it's in my "path" variable 
Is this normal or can it cause issues?

**Reply by ossirytk** [reply]
I think the issue here is that the set for environment variables does not actually set them. So you don't actually build with cuda support enabled. This has been commented here somewhere and in the llama-cpp-python issues. I can't seem to find the issue where I wrote comments and it might have been cleaned already, but I vaguely remember that I use the powershell `$Env:<variable-name> = "<new-value>"` syntax to enable cuda. That or this needs to be `set CMAKE_ARGS="-DLLAMA_CUBLAS=on"`(Edit: This seems to be the syntax in the current readme, but it did not set the value for me in powershell). Just looking at your set CMAKE_ARG line, it seems wrong. You can check the content with `$Env:Foo` to make sure that it was set correctly.

**Reply by ossirytk** [reply]
I think I found the issue #2625

**Reply by viperwasp** [reply]
I will check into this when I can. I copy and pasted the entire page including the more detailed information page into chat GTP and asked what code to use. I did not do this right away but I wanted to double check a few things. 

I tried this following code and same results issue not fixed. But maybe this code is wrong too. 

pip uninstall -y llama-cpp-python
set LLAMA_CUBLAS=1
set CMAKE_ARGS="-DLLAMA_CUBLAS=on"
set FORCE_CMAKE=1
pip install llama-cpp-python --no-cache-dir

You also have to run each line one at a time to avoid error at least in CMD. 

Thanks though ossirytk seems like something is up. I won't be working on this until later today so maybe someone will know for sure the exact code to run and if it's better to use powershell or not. Probably won't matter if using correct code.

**Reply by ossirytk** [reply]
I don't think you need `set LLAMA_CUBLAS=1` It should be enough to run
`set CMAKE_ARGS="-DLLAMA_CUBLAS=on"
set FORCE_CMAKE=1
pip install llama-cpp-python --upgrade --force-reinstall --no-cache-dir`

**Comment by viperwasp** [comment]
Thanks I will try that code. Do I have to completely delete and put in Ooba again or can I just run that code now? 
Also if I don't get it working in next 48 hours likely going to try the WSL2 however knowing my luck it won't work either. lol

**Reply by viperwasp** [reply]
Ran code here is what I got...

C:\AI\oobabooga_windows2>set FORCE_CMAKE=1

C:\AI\oobabooga_windows2>set CMAKE_ARGS="-DLLAMA_CUBLAS=on"

C:\AI\oobabooga_windows2>pip install llama-cpp-python --upgrade --force-reinstall --no-cache-dir
Collecting llama-cpp-python
  Downloading llama_cpp_python-0.1.62.tar.gz (1.4 MB)
     ---------------------------------------- 1.4/1.4 MB 23.2 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting typing-extensions>=4.5.0 (from llama-cpp-python)
  Downloading typing_extensions-4.6.3-py3-none-any.whl (31 kB)
Collecting numpy>=1.20.0 (from llama-cpp-python)
  Downloading numpy-1.24.3-cp310-cp310-win_amd64.whl (14.8 MB)
     ---------------------------------------- 14.8/14.8 MB 72.5 MB/s eta 0:00:00
Collecting diskcache>=5.6.1 (from llama-cpp-python)
  Downloading diskcache-5.6.1-py3-none-any.whl (45 kB)
     ---------------------------------------- 45.6/45.6 kB ? eta 0:00:00
Building wheels for collected packages: llama-cpp-python
  Building wheel for llama-cpp-python (pyproject.toml) ... done
  Created wheel for llama-cpp-python: filename=llama_cpp_python-0.1.62-cp310-cp310-win_amd64.whl size=534247 sha256=f4bf0a36b7d4c16f1881522c7e80355c04362456d860453721a945bb2a4ab297
  Stored in directory: C:\Users\Johndoe\AppData\Local\Temp\pip-ephem-wheel-cache-74ge78l6\wheels\70\ff\8a\a05628c7dc164b9d74b02168b42ae7436c7159cd8d714b57fe
Successfully built llama-cpp-python
Installing collected packages: typing-extensions, numpy, diskcache, llama-cpp-python
  Attempting uninstall: typing-extensions
    Found existing installation: typing_extensions 4.6.3
    Uninstalling typing_extensions-4.6.3:
      Successfully uninstalled typing_extensions-4.6.3
  Attempting uninstall: numpy
    Found existing installation: numpy 1.24.3
    Uninstalling numpy-1.24.3:
      Successfully uninstalled numpy-1.24.3
  Attempting uninstall: diskcache
    Found existing installation: diskcache 5.6.1
    Uninstalling diskcache-5.6.1:
      Successfully uninstalled diskcache-5.6.1
  Attempting uninstall: llama-cpp-python
    Found existing installation: llama-cpp-python 0.1.62
    Uninstalling llama-cpp-python-0.1.62:
      Successfully uninstalled llama-cpp-python-0.1.62
**ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
xformers 0.0.20 requires torch==2.0.1, but you have torch 2.1.0.dev20230612+cu121 which is incompatible.
Successfully installed diskcache-5.6.1 llama-cpp-python-0.1.62 numpy-1.24.3 typing-extensions-4.6.3**

I think chat GTP suggested "pip install xformers --upgrade" but that gives me a different version of pytorch without cuda for some reasons. I'm not sure how important this error is though. But if you see anything else off let me know. And no it's still not working.

**Reply by ossirytk** [reply]
You should resolve the pip conflicts first. Maybe get rid of anything that's not part of the vanilla installation. Uninstalling first might be a good idea, sice you have conflicts. Everything "Optional" or "Recommended" should go. Then you should test that the env variables were set correctly. If the variables were set and pip resolves, you should be good.

**Reply by viperwasp** [reply]
That's the "issue" though I was told to put in latest Nvidia Cuda Toolkit. Which happens to be cu121. I need pytorch that can run cu121 right? https://pytorch.org/ I have the only version that says 12.1 I think the only reason I upgraded my pytorch in the first place is a more vanilla setup did not work at all. 

I've tried so many different things. But if I need to put in a different Nvida toolkit like cuda 117 I could try that. If I can't upgrade xformers and I can't install a version of Pytorch that works and is cu121 then I don't know what else I can do.

**Reply by ossirytk** [reply]
I don't think you need the pytorch to just run the chat. I installed pytorch, but it doesn't match the latest cuda toolkit version I installed either. I did not install any extra things. Pytorch might be required if you compile the model binary yourself. The same with bitwise. Pretty sure you don't need it and can just ignore the errors if it complains about not being compiled with gpu support.

**Reply by viperwasp** [reply]
Okay so in that case than I still think I am stuck. I am 99% sure I did all of this already. Like I only just recently updated Pytorch was like attempt number 14! Just earlier today I had a version and I was getting zero errors. I combiled the GPU accleration branch of llama.cpp. Now then again maybe there was that coding issue with value not updating. However I did it many different way. But that will be my next job. Put in pytorch that has no errors. Keep Nvidia Cuda 12.1. Make sure I run the right commands for the llama.cpp and I am 99% sure I've done that. Thank you.

**Reply by ossirytk** [reply]
How exactly are you installing everything? Just for the point of reference I'm running everything in conda virtual env and I usually have up to date develop. I just run updates from the requirements.txt.

**Reply by viperwasp** [reply]
One click installer and then CMD using pip commands mostly. I don't know how to use Conda. I have miniconda installed I think. I've never really used those in any way shape or form except maybe when setting up Stable Diffusion or something but I think it was an old way of doing that. 

By the way I just put in a different Pytorch. Used the updated commands for blas. No errors during any step. Still not working. I assume I have the settings correct. I'm not using any arguments by the way. I just launch the model via "start_windows.bat"
![Capture](https://github.com/oobabooga/text-generation-webui/assets/38386267/a7d670c3-05fd-42b6-997d-db6e6189c44e)

**Reply by ossirytk** [reply, marked as answer]
This look roughly correct. Does the model load? An is the gpu use ~0%? If that is the case, then the env variables were probably set incorrectly. In that case you'll want to try setting the variables first and check that the were correct. I don't remember the cmd syntax for this, but you can google something like "print env variable in command line" or something like that. If it prints empty or something nonsense, then the set was incorrect. like I mentioned in the task I posted, I used powershell and the commands I posted to enable cuda.

**Reply by viperwasp** [reply]
I think they are being set right? 
C:\AI\oobabooga_windows2>echo %CMAKE_ARGS%
%CMAKE_ARGS%

C:\AI\oobabooga_windows2>set CMAKE_ARGS="-DLLAMA_CUBLAS=on" && set FORCE_CMAKE=1

C:\AI\oobabooga_windows2>echo %CMAKE_ARGS%
"-DLLAMA_CUBLAS=on"

C:\AI\oobabooga_windows2>echo %FORCE_CMAKE%
1

I ran pip install llama-cpp-python --upgrade --force-reinstall --no-cache-dir 
the next line after. It completed. No error. Still same issue. Blas=0 and my Video ram stays at 10% same  not changing when I load a model etc.

**Reply by viperwasp** [reply]
One question though? When I do pip uninstall -y llama-cpp-python it still runs GGML models? Is this unexpected? Am I like installing to a wrong environment or something? Or will ooba load GGML to CPU/Ram without llama-cpp-python? I think for some reason I need to use conda to update llama or something? 

Update
Okay I am learning how to use Anaconda Prompt Miniconda3 and I think I can also use it via CMD too by going into the right envirorments I just learned that two in my oobabooa folder env and conda.  My gut tells me that the env one is the one I need to access and do the llama.cpp with GPU acceleration there. I think I have been doing it wrong. So I need to tell CMD to link to this env somehow. Might take me a bit.

**Reply by viperwasp** [reply]
Well there we go! It's "fixed" all this time the main issue was that I was not updating this in the env. I was just using pip within regular cmd. Which many times worked fine. I got errors in ooba saying missing a pip dependency I would open cmd and do a pip command and download the missing package etc. I did it that way for many things! But for whatever reason this was different. I feel stupid but at the same time I did not know that environments existed until today. Brand new concept to me.  

To be fair this was not my only problem. Visual Studios was also properly missing from SYSTEM "PATH" so that would have stopped me in my tracks too. Likely a few other small issues I learned and fixed though all of this. 

Thanks everyone who has helped ossirytk, jparmstr and honestly several people across many forums.  

It seems to be working already did a test prompt. 

llama_model_load_internal: using CUDA for GPU acceleration
llama_model_load_internal: mem required  = 12580.00 MB (+ 3124.00 MB per state)
llama_model_load_internal: allocating batch_size x 1 MB = 512 MB VRAM for the scratch buffer
llama_model_load_internal: offloading 25 layers to GPU
llama_model_load_internal: total VRAM used: 7689 MB
...........................................
llama_init_from_file: kv self size  = 3120.00 MB
AVX = 1 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | VSX = 0 |

Output generated in 6.42 seconds (1.40 tokens/s, 9 tokens, context 41, seed 744249115)

I just need to allocate the optimal amount of VRAM this was just a test. But these are the things I'm actually good at.

**Reply by ossirytk** [reply]
Time to pop open some champagne.
